packages/vite-plugin/src/index.ts

```diff
@@ -1,62 +1,98 @@
-import outdent from 'outdent';
+import type { Plugin, ResolvedConfig, ViteDevServer, Rollup } from 'vite';
 import {
   cssFileFilter,
-  processVanillaFile,
-  compile,
   IdentifierOption,
   getPackageInfo,
-  CompileOptions,
   transform,
+  type Compiler,
+  createCompiler,
+  normalizePath,
 } from '@vanilla-extract/integration';
-import { type PostCSSConfigResult, resolvePostcssConfig } from './postcss';
-
-const styleUpdateEvent = (fileId: string) =>
-  `vanilla-extract-style-update:${fileId}`;
 
 const virtualExtCss = '.vanilla.css';
-const virtualExtJs = '.vanilla.js';
+
+const isVirtualId = (id: string) => id.endsWith(virtualExtCss);
+const fileIdToVirtualId = (id: string) => `${id}${virtualExtCss}`;
+const virtualIdToFileId = (virtualId: string) =>
+  virtualId.slice(0, -virtualExtCss.length);
 
 interface Options {
   identifiers?: IdentifierOption;
-  emitCssInSsr?: boolean;
-  esbuildOptions?: CompileOptions['esbuildOptions'];
+  unstable_mode?: 'transform' | 'emitCss';
 }
 export function vanillaExtractPlugin({
   identifiers,
-  emitCssInSsr,
-  esbuildOptions,
+  unstable_mode: mode = 'emitCss',
 }: Options = {}): Plugin {
   let config: ResolvedConfig;
   let server: ViteDevServer;
-  let postCssConfig: PostCSSConfigResult | null;
-  // We lazily load this utility from Vite
-  let normalizePath: (fsPath: string) => string;
-  const cssMap = new Map<string, string>();
-
-  const hasEmitCssOverride = typeof emitCssInSsr === 'boolean';
-  let resolvedEmitCssInSsr: boolean = hasEmitCssOverride
-    ? emitCssInSsr
-    : !!process.env.VITE_RSC_BUILD;
   let packageName: string;
+  let compiler: Compiler | undefined;
+
+  const getIdentOption = () =>
+    identifiers ?? (config.mode === 'production' ? 'short' : 'debug');
+  const getAbsoluteId = (filePath: string) => {
+    let resolvedId = filePath;
+
+    if (
+      filePath.startsWith(config.root) ||
+      // In monorepos the absolute path will be outside of config.root, so we check that they have the same root on the file system
+      (path.isAbsolute(filePath) &&
+        filePath.split(path.sep)[1] === config.root.split(path.sep)[1])
+    ) {
+      resolvedId = filePath;
+    } else {
+      // In SSR mode we can have paths like /app/styles.css.ts
+      resolvedId = path.join(config.root, filePath);
+    }
+
+    return normalizePath(resolvedId);
+  };
 
-  const getAbsoluteVirtualFileId = (source: string) =>
-    normalizePath(path.join(config.root, source));
+  function invalidateModule(absoluteId: string) {
+    if (!server) return;
+
+    const { moduleGraph } = server;
+    const modules = Array.from(moduleGraph.getModulesByFile(absoluteId) || []);
+
+    for (const module of modules) {
+      moduleGraph.invalidateModule(module);
+
+      // Vite uses this timestamp to add `?t=` query string automatically for HMR.
+      module.lastHMRTimestamp = module.lastInvalidationTimestamp || Date.now();
+    }
```

Comment:
You could keep this as a `Set` and avoid the extra iteration from `Array.from`:

```suggestion
    const modules = moduleGraph.getModulesByFile(absoluteId);
    
    if (modules) { 
      for (const module of modules) {
        moduleGraph.invalidateModule(module);
  
        // Vite uses this timestamp to add `?t=` query string automatically for HMR.
        module.lastHMRTimestamp = module.lastInvalidationTimestamp || Date.now();
      }
    }
```